### ISSUE TITLE
Replace Viper with a config struct

### ISSUE THREAD

**Issue by sagikazarmark** [dunglas/mercure#366]
Generally speaking, Viper is not supposed to be used outside of the main function. It's job is to collect configuration and then pass it onto a config struct/structs, something like this:

```
type Config struct {
	Debug                  bool          `mapstructure:"debug"`
	TransportURL           string        `mapstructure:"transport_url"`
	JWTKey                 string        `mapstructure:"jwt_key"`
	JWTAlgorithm           string        `mapstructure:"jwt_algorithm"`
	PublisherJWTKey        string        `mapstructure:"publisher_jwt_key"`
	PublisherJWTAlgorithm  string        `mapstructure:"publisher_jwt_algorithm"`
	SubscriberJWTKey       string        `mapstructure:"subscriber_jwt_key"`
	SubscriberJWTAlgorithm string        `mapstructure:"subscriber_jwt_algorithm"`
	AllowAnonymous         bool          `mapstructure:"allow_anonymous"`
	CORSAllowedOrigins     []string      `mapstructure:"cors_allowed_origins"`
	PublishAllowedOrigins  []string      `mapstructure:"publish_allowed_origins"`
	Addr                   string        `mapstructure:"addr"`
	AcmeHTTP01Addr         string        `mapstructure:"acme_http01_addr"`
	AcmeHosts              []string      `mapstructure:"acme_hosts"`
	AcmeCertDir            string        `mapstructure:"acne_cert_dir"`
	CertFile               string        `mapstructure:"cert_file"`
	KeyFile                string        `mapstructure:"key_file"`
	HeartbeatInterval      time.Duration `mapstructure:"heartbeat_interval"`
	ReadTimeout            time.Duration `mapstructure:"read_timeout"`
	WriteTimeout           time.Duration `mapstructure:"write_timeout"`
	DispatchTimeout        time.Duration `mapstructure:"dispatch_timeout"`
	Compress               bool          `mapstructure:"compress"`
	UseForwardedHeaders    bool          `mapstructure:"use_forwarded_headers"`
	Demo                   bool          `mapstructure:"demo"`
	LogFormat              string        `mapstructure:"log_format"`
	Subscriptions          bool          `mapstructure:"subscriptions"`
	Metrics                bool          `mapstructure:"metrics"`
	MetricsLogin           string        `mapstructure:"metrics_login"`
	MetricsPassword        string        `mapstructure:"metrics_password"`
}
```

It separates your code nicely from the CLI/Viper stuff and makes testing easier. I can provide a PR if you like the idea.

**Comment by dunglas** [comment]
Actually, it's how it worked before https://github.com/dunglas/mercure/pull/203.

I'm curious of why Viper shouldn't be used outside of the main function? Using it as a config registry looks legit, and allows to leverage easily nice features such as live reloading. Viper is described like this on its homepage ("a registry for all your application configuration needs").
Hugo and doctl use Viper like this too.

That being said, I was already considering switching back to a plain old config struct (using https://github.com/spf13/viper#unmarshaling as a first step) because I plan to make Mercure also available both as a standalone lib and as a Caddy module (https://caddyserver.com/docs/extending-caddy). Caddy has some requirements regarding the config: https://caddyserver.com/docs/extending-caddy#configuration
Basically, the config struct must be compatible with JSON.

The next step for Caddy compatibility will be to switch from Logrus to Zap (https://caddyserver.com/docs/extending-caddy#logs).

WDYT?

**Comment by sagikazarmark** [comment]
There are several issues with Viper as it is. You can read more about them here: https://github.com/spf13/viper/projects/1

But basically the main problem is Viper leaking a bunch of runtime responsibilities into your code, making it harder to test. Whereas with a single struct the configuration is more readable, and the whole config loading and processing thing is limited to your `main` function, where it should be anyway.

Ever since unmarshaling is a thing, I recommend everyone to use it instead of relying on viper directly in business logic.

You can still utilize live reloading, you just need to unmarshal the config again and inject it into your components.

> because I plan to make Mercure also available both as a standalone lib

Huge :+1:

I actually had to fork Mercure because I needed a couple extra features (will provide PRs when I get there), but it'd be so much better if I could just use it as a library, replacing the parts I need.

> Basically, the config struct must be compatible with JSON.

I don't think this is a problem, but I would probably not compromise the app config here. If you need to, you can create a separate config layer for Caddy and "convert" it to your config format/struct. That way you are not tied to Caddy. It's a bit more work though.

The alternative is keeping the two config the same, which has the advantage of not having to document both.

> The next step for Caddy compatibility will be to switch from Logrus to Zap

Again, I would suggest avoiding a vendor lock-in. You could define a logger interface for Mercure and provide a default implementation for Zap. That way your users can choose whatever logger they want to use when using it as a library and you spare them the extra dependency.

Take a look at https://github.com/logur/logur which is an example for such interface. You don't actually have to import it into your project, you can just copy the interface, keep the relevant levels, etc. If you remain compatible with the original interface though, your users will be able to use the adapters that come with the project, so they don't actually have to implement your interface.

I like those plans overall, I'd be happy to help executing them.

**Comment by sagikazarmark** [comment]
I was exploring the code and had another idea: by the looks of it, the primary business logic for the code is not actually tied to the HTTP layer. I mean, there are obviously HTTP specific things in there (like sending specific headers), but most of the logic is not HTTP specific.

I was wondering if it would make sense to separate some parts of the code (eg. accept a `WriteFlusher` interface, instead of an `http.Response`, load form values upfront and pass a struct to the main logic) from the HTTP layer.

The obvious benefit is better tests for the business logic (e2e tests can still cover the whole thing).
